WatsonWebsocket/WatsonWsClient.cs

```diff
@@ -1,472 +1,813 @@
+            {
+                DisconnectResetSubscriptions();
+            }
+            _ClientWs.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, "", CancellationToken.None).Wait();
+        }
+
+        /// <summary>
+        /// Send text data to the server asynchronously.
+        /// </summary>
+        /// <param name="data">String data.</param>
+        /// <param name="token">Cancellation token allowing for termination of this request.</param>
+        /// <returns>Task with Boolean indicating if the message was sent successfully.</returns>
+        public async Task<bool> SendAsync(string data, CancellationToken token = default)
+        {
+            if (String.IsNullOrEmpty(data)) throw new ArgumentNullException(nameof(data));
+            else return await MessageWriteAsync(Encoding.UTF8.GetBytes(data), WebSocketMessageType.Text, token);
+        }
+
+        /// <summary>
+        /// Send binary data to the server asynchronously.
+        /// </summary>
+        /// <param name="data">Byte array containing data.</param>
+        /// <param name="token">Cancellation token allowing for termination of this request.</param>
+        /// <returns>Task with Boolean indicating if the message was sent successfully.</returns>
+        public async Task<bool> SendAsync(byte[] data, CancellationToken token = default)
+        {
+            if (data == null || data.Length < 1) throw new ArgumentNullException(nameof(data));
+            return await MessageWriteAsync(data, WebSocketMessageType.Binary, token);
+        }
+
+        /// <summary>
+        /// Send text data to the server asynchronously and wait for anything response.
+        /// </summary>
+        /// <param name="data">String data.</param>
+        /// <param name="timeout">Maximum waiting time until one response.</param>
+        /// <param name="token">Cancellation token allowing for termination of this request.</param>
+        /// <returns>Task with Boolean indicating if the message was sent successfully.</returns>
+        public async Task<string> SendAndReplyAnythingAsync(string data, int timeout = 30, CancellationToken token = default)
+        {
+            if (data == null || data.Length < 1) throw new ArgumentNullException(nameof(data));
+            if (timeout < 1) throw new ArgumentException("Timeout invalid", nameof(data));
+            string result = null;
+
+            await Task.Run(async () =>
+            {
+                // Create Catcher Data
+                var receivedEvent = new ManualResetEvent(false);
+                InternalMessageReceived += (s, e) =>
+                {
+                    result = Encoding.UTF8.GetString(e.Data);
+                    receivedEvent.Set();
+                };
+                // Detect Connection Lost
+                InternalServerDisconnected += (s, e) => receivedEvent.Set();
+
+                await MessageWriteAsync(Encoding.UTF8.GetBytes(data), WebSocketMessageType.Binary, token);

```

Comment:
I don't think this can guarantee that the next incoming message is actually a response to the one that is sent.  There needs to be some kind of correlating ID that is sent with the outgoing message, that the recipient should also include in the response.

Reply:
Totally, but that is left to the programmer who uses the library since he must know what data he sends and receives. For example, I implemented ProtoBuf on this library and it was specified which message should arrive and which one to expect.